### PR TITLE
Fix hash modification while iterating

### PR DESCRIPTION
**Pull request by jgraichen**
Fixes a rare concurrent hash access error by replacing the IO event loop with a new implementation and improving the logger receiver. See commits for more.

### CODE REVIEW

**Review by franzliedke — APPROVED** [review]
Looking sensible and easier to understand than the old implementation, but I cannot claim that I know about how `nio4r` works exactly. ;)

**Comment by jgraichen** [comment]
It is surprisingly easy. You might want to read https://github.com/socketry/nio4r/wiki.

File: lib/multi_process/loop.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+require 'nio'
+
+module MultiProcess
+  class Loop
+    def initialize
+      @selector = ::NIO::Selector.new
+      @queue = ::Queue.new
```
Comment (mswart) [inline lib/multi_process/loop.rb]: Isn't this unused?

File: lib/multi_process/loop.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+          sleep(0.001)
+        end
+      end
+    end
+
+    def watch(io, &block)
+      @selector.wakeup
+      @selector.register(io, :r).tap do |monitor|
+        monitor.value = block
+        block.call(:registered, monitor)
```
Comment (mswart) [inline lib/multi_process/loop.rb]: Maybe use `monitor.value.call` for consistence.

**Review by mswart — APPROVED** [review]
Looks good.

**Pull request merged**